### PR TITLE
Testcase for adding and validating backup rule

### PR DESCRIPTION
**Pull request by snigdha-px**
**What this PR does / why we need it**:
- Deploy and validate application
- Create backup pre and post rule
- Validate rule
**Which issue(s) this PR fixes** (optional)
Closes # PA-260, PA-264

**Special notes for your reviewer**:
Result: TestSetId 5772 update successfully
http://aetos.pwx.purestorage.com/resultSet/testSetID/5772

### CODE REVIEW

File: tests/backup/backup_basic_test.go [hunk 1]
```diff
@@ -45,6 +57,7 @@ func TestBasic(t *testing.T) {
 }
 
 var _ = BeforeSuite(func() {
+	dash = Inst().Dash
 	logrus.Infof("Init instance")
```
Comment (lsrinivas-pure) [inline tests/backup/backup_basic_test.go:61]: Add the logger change

File: tests/backup/backup_test.go [hunk 2]
```diff
@@ -60,14 +57,121 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
...
+		log.Infof(" No pre-setup required for this testcase")
+	})
 	It("Backup Cluster Verification", func() {
-		Step("Check the status of pxcentral-post-install-hook pods", func() {
+		Step("Check the status of backup pods", func() {
+			dash.Info("Check the status of backup pods")
 			status := validateBackupCluster()
-			Expect(status).NotTo(Equal(false),
-				fmt.Sprintf("Backup pods are not in expected state"))
+			dash.VerifyFatal(status, false, fmt.Sprintf("Backup pods are not in expected state"))
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:67]: Fix the fatal condition, I think true is expected here

File: tests/backup/backup_basic_test.go [hunk 3]
```diff
@@ -45,6 +57,7 @@ func TestBasic(t *testing.T) {
...
 	InitInstance()
```
Comment (lsrinivas-pure) [inline tests/backup/backup_basic_test.go:62]: Do testset begin here

File: tests/backup/backup_test.go [hunk 4]
```diff
@@ -60,14 +57,121 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
 
 //This testcase verifies if the backup pods are in Ready state or not
 var _ = Describe("{BackupClusterVerification}", func() {
+	JustBeforeEach(func() {
+		log.Infof(" No pre-setup required for this testcase")
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:61]: you need to do testcase begin here else dash messages will not be posted

File: tests/backup/backup_test.go (hunk 2 above)
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:67]: Fix the verifircation description like Validating Backup pods

File: tests/backup/backup_test.go [hunk 5]
```diff
@@ -60,14 +57,121 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
...
+			dash.Info("Check the status of backup pods")
 			status := validateBackupCluster()
-			Expect(status).NotTo(Equal(false),
-				fmt.Sprintf("Backup pods are not in expected state"))
+			dash.VerifyFatal(status, false, fmt.Sprintf("Backup pods are not in expected state"))
 		})
 		//Will add CRD verification here
 	})
+	JustAfterEach(func() {
+		log.Infof(" No cleanup required for this testcase")
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:72]: Do testcaseend after test is completed

File: tests/backup/backup_test.go [hunk 6]
```diff
@@ -60,14 +57,121 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
...
+			}
+		})
+	})
+	JustAfterEach(func() {
+		dash.Info("Deleting the deployed apps after the testcase")
+		for i := 0; i < Inst().GlobalScaleFactor; i++ {
+			opts := make(map[string]bool)
+			opts[SkipClusterScopedObjects] = true
+			taskName := fmt.Sprintf("%s-%d", taskNamePrefix, i)
+			err := Inst().S.Destroy(contexts[i], opts)
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:171]: Add testcaseend here once all tests are completed

File: tests/backup/backup_test.go [hunk 7]
```diff
@@ -2008,3 +2112,68 @@ func validateBackupCluster() bool {
 	}
 	return true
 }
+
+func createRuleForBackup(rule_name string, orgID string, pod_selector []map[string]string, action_value []string,
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:2116]: create common file and move this function to that

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

File: tests/backup/backup_basic_test.go (hunk 1 above)
Comment (snigdha-px) [reply tests/backup/backup_basic_test.go:61]: done!!

File: tests/backup/backup_test.go (hunk 2 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:67]: done!!

File: tests/backup/backup_basic_test.go (hunk 3 above)
Comment (snigdha-px) [reply tests/backup/backup_basic_test.go:62]: done!!

File: tests/backup/backup_test.go (hunk 5 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:72]: done!!

File: tests/backup/backup_test.go (hunk 6 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:171]: done!!

File: tests/backup/common.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+		Uid:   uid,
+	}
+	_, err1 := Inst().Backup.InspectRule(ctx, RuleInspectReq)
+	if err1 != nil {
+		logrus.Errorf("Failed to validate the created rule with Error: [%v]", err)
+		return false
+	}
+	return true
+}
+func contains(s []string, str string) bool {
```
Comment (vpinisetti-px) [inline tests/backup/common.go:135]: better to give some meaningful name instead of "s" and "v"

File: tests/backup/backup_test.go [hunk 9]
```diff
@@ -60,14 +55,125 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
...
+		dash.Info("Deploy applications")
+		dash.TestCaseBegin("Backup: BasicBackupCreateWithRules", "Creating backup with Rules", "", nil)
+		contexts = make([]*scheduler.Context, 0)
+		for i := 0; i < Inst().GlobalScaleFactor; i++ {
+			taskName := fmt.Sprintf("%s-%d", taskNamePrefix, i)
+			appContexts = ScheduleApplications(taskName)
+			contexts = append(contexts, appContexts...)
+		}
+	})
+	It("Baic Backup Creation", func() {
```
Comment (kshithijiyer-px) [inline tests/backup/backup_test.go:90]: Typo: Basic

File: tests/backup/backup_basic_test.go [hunk 10]
```diff
@@ -33,8 +33,15 @@ var (
 	// selectively enable after suite actions by setting wantAllAfterSuiteActions to false and setting these to true
 	wantAfterSuiteSystemCheck     bool = false
 	wantAfterSuiteValidateCleanup bool = false
+	pre_action_list                    = map[string]string{"cassandra": "nodetool flush -- keyspace1;", "postgres": "PGPASSWORD=$POSTGRES_PASSWORD; psql -U '$POSTGRES_USER' -c 'CHECKPOINT';"}
```
Comment (kshithijiyer-px) [inline tests/backup/backup_basic_test.go:36]: Can we use a nested map here? 

```
	currency := map[string]map[string]int{
		"cassandra": {"pre_action_list": 'action', 'post_action_list': 'action'},
	}

```

File: tests/backup/backup_test.go [hunk 11]
```diff
@@ -60,14 +55,125 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
 
 //This testcase verifies if the backup pods are in Ready state or not
 var _ = Describe("{BackupClusterVerification}", func() {
+	JustBeforeEach(func() {
+		log.Infof(" No pre-setup required for this testcase")
```
Comment (kshithijiyer-px) [inline tests/backup/backup_test.go:59]: Unnecessary space at start of log line.

File: tests/backup/backup_test.go [hunk 12]
```diff
@@ -60,14 +55,125 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
...
+					pre_container_list = append(pre_container_list, os.Getenv(container_name))
+				}
+			}
+			if create_pre_rule == true {
+				pre_rule_status := createRuleForBackup("backup-pre-rule", "default", pre_ps_list, pre_actions,
+					pre_background, pre_run_in_single_pod, pre_container_list)
+				fmt.Println(pre_rule_status)
+				dash.VerifyFatal(pre_rule_status, true, fmt.Sprintf("Verifying pre rule for backup"))
+			}
+			dash.Info(" Creating post rule for deployed apps")
```
Comment (kshithijiyer-px) [inline tests/backup/backup_test.go:140]: Unnecessary space at the start of the log line.

File: tests/backup/backup_test.go [hunk 13]
```diff
@@ -60,14 +55,125 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
...
+			for i := 0; i < len(app_list); i++ {
+				if app_list[i] == "cassandra" {
+					create_post_rule = true
+					post_ps_list = append(post_ps_list, ps[app_list[i]])
+					post_actions = append(post_actions, post_action_list[app_list[i]])
+					post_background = append(post_background, background_check_dict[app_list[i]])
+					post_run_in_single_pod = append(post_run_in_single_pod, run_in_single_pod_dict[app_list[i]])
+					container_name := fmt.Sprintf("%s%s", "container", app_list[i])
+					fmt.Println(" Conainer name is ", container_name)
+					post_container_list = append(post_container_list, os.Getenv(container_name))
```
Comment (kshithijiyer-px) [inline tests/backup/backup_test.go:155]: Why do we need to store it in a variable can't we directly pass it in the below function?

File: tests/backup/common.go [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+			equal1, _ := regexp.MatchString(full_maintenance_pod, pod.GetName())
+			if !(equal || equal1) {
+				logrus.Info("Checking if all the containers are up or not")
+				res := core.Instance().IsPodRunning(pod)
+				if !res {
+					logrus.Errorf("All the containers are not Up")
+					return false
+				}
+				err = core.Instance().ValidatePod(&pod, defaultTimeout, defaultTimeout)
+				logrus.Warnf(" ERR is %s", err)
```
Comment (kshithijiyer-px) [inline tests/backup/common.go:68]: Unnecessary  space in from of log line.

**Review by kshithijiyer-px — CHANGES_REQUESTED** [review]

File: tests/backup/backup_test.go (hunk 7 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go]: done

File: tests/backup/common.go (hunk 8 above)
Comment (snigdha-px) [reply tests/backup/common.go:135]: done!!

File: tests/backup/backup_basic_test.go (hunk 10 above)
Comment (snigdha-px) [reply tests/backup/backup_basic_test.go:36]: done!!

File: tests/backup/backup_test.go (hunk 11 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:59]: done!!

File: tests/backup/backup_test.go (hunk 12 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:140]: done!!

File: tests/backup/common.go (hunk 14 above)
Comment (snigdha-px) [reply tests/backup/common.go:68]: done!!

File: tests/backup/backup_test.go (hunk 9 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:90]: done!

File: tests/backup/backup_test.go [hunk 15]
```diff
@@ -60,14 +55,125 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
...
+					} else if obj, ok := specObj.(*appsapi.StatefulSet); ok {
+						if contains(app_list, obj.Name) == true {
+							ps[obj.Name] = obj.Spec.Template.Labels
+						}
+					}
+				}
+			}
+		})
+		Step("Creating rules for backup", func() {
+			pre_ps_list := []map[string]string{}
```
[the remark this answers is not in the document]
Comment (snigdha-px) [reply tests/backup/backup_test.go:115]: done!

File: tests/backup/backup_test.go (hunk 13 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:155]: done!

**Review by kshithijiyer-px — APPROVED** [review]
LGTM

File: tests/backup/backup_test.go [hunk 16]
```diff
@@ -60,14 +54,136 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
...
+				time.Sleep(time.Minute * 1)
+				CreateBackupLocation(provider, backup_location_name, BackupLocationUID, CredName, CloudCredUID, bucketName, orgID)
+			}
+		})
+	})
+	JustAfterEach(func() {
+		dash.Info("Deleting the deployed apps after the testcase")
+		for i := 0; i < len(contexts); i++ {
+			opts := make(map[string]bool)
+			opts[SkipClusterScopedObjects] = true
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:143]: As discussed make this teardown as a common method

File: tests/backup/backup_test.go [hunk 17]
```diff
@@ -60,14 +54,136 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
...
+			taskName := fmt.Sprintf("%s-%d", taskNamePrefix, i)
+			err := Inst().S.Destroy(contexts[i], opts)
+			dash.VerifySafely(err, nil, fmt.Sprintf("Verify destroying app %s, Err: %v", taskName, err))
+		}
+		dash.Info("Deleting backup rules created")
+		RuleEnumerateReq := &api.RuleEnumerateRequest{
+			OrgId: orgID,
+		}
+		ctx, err := backup.GetAdminCtxFromSecret()
+		Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Failed to fetch px-central-admin ctx: [%v]", err))
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:153]: use dash.VerifyFatal every where you have Expect

File: tests/backup/common.go [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+	labelSelectors := map[string]string{"job-name": post_install_hook_pod}
+	ns := backup.GetPxBackupNamespace()
+	pods, err := core.Instance().GetPods(ns, labelSelectors)
+	if err != nil {
+		logrus.Errorf("Unable to fetch pxcentral-post-install-hook pod from backup namespace\n Error : [%v]\n",
+			err)
+		return false
+	}
+	for _, pod := range pods.Items {
+		logrus.Info("Checking if the pxcentral-post-install-hook pod is in Completed state or not")
```
Comment (lsrinivas-pure) [inline tests/backup/common.go:49]: change to log object

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

File: tests/backup/common.go [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+				pod_selector = append(pod_selector, ps[app_list[i]])
+				action_value = append(action_value, app_parameters[app_list[i]]["post_action_list"])
+				background_val, _ := strconv.ParseBool(app_parameters[app_list[i]]["background"])
+				background = append(background, background_val)
+				pod_val, _ := strconv.ParseBool(app_parameters[app_list[i]]["run_in_single_pod"])
+				run_in_single_pod = append(run_in_single_pod, pod_val)
+				// Here user has to set env for each app container if required in the format container<app name> eg: containersql
+				container_name := fmt.Sprintf("%s-%s", "container", app_list[i])
+				container = append(container, os.Getenv(container_name))
+			}
```
Comment (vpinisetti-px) [inline tests/backup/common.go:157]: @snigdha-px will have to handle the error incase requested app is not found in the list.

**Review by vpinisetti-px — CHANGES_REQUESTED** [review]

File: tests/backup/backup_test.go (hunk 17 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:153]: Done!!

File: tests/backup/common.go (hunk 18 above)
Comment (snigdha-px) [reply tests/backup/common.go:49]: Done!!

File: tests/backup/backup_test.go (hunk 16 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:143]: Done!!

File: tests/backup/common.go (hunk 19 above)
Comment (snigdha-px) [reply tests/backup/common.go:157]: done

File: tests/backup/backup_basic_test.go [hunk 20]
```diff
@@ -56,6 +57,7 @@ var _ = AfterSuite(func() {
 	if wantAllAfterSuiteActions || wantAfterSuiteValidateCleanup {
 		ValidateCleanup()
 	}
+	defer dash.TestCaseEnd()
```
Comment (lsrinivas-pure) [inline tests/backup/backup_basic_test.go:60]: Defer should be the first statement,
Also add TestSetEnd

File: tests/backup/backup_test.go [hunk 21]
```diff
@@ -60,14 +54,107 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
...
+				}
+			}
+			if contains(post_rule_app, app_list[i]) == true {
+				if _, ok := app_parameters[app_list[i]]["post_action_list"]; ok {
+					dash.VerifyFatal(ok, true, fmt.Sprintf("Pre Rule details mentioned for the apps"))
+				}
+			}
+		}
+		dash.Info("Deploy applications")
+		dash.TestCaseBegin("Backup: BasicBackupCreateWithRules", "Creating backup with Rules", "", nil)
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:101]: TestCaseBegin should be first statement

File: tests/backup/backup_test.go [hunk 22]
```diff
@@ -60,14 +54,107 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
...
+		}
+	})
+	It("Basic Backup Creation", func() {
+		Step("Validate applications and get their labels", func() {
+			ValidateApplications(contexts)
+			log.Infof("Create list of pod selector for the apps deployed")
+			for _, ctx := range appContexts {
+				for _, specObj := range ctx.App.SpecList {
+					if obj, ok := specObj.(*appsapi.Deployment); ok {
+						if contains(app_list, obj.Name) == true {
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:116]: no need for == here

File: tests/backup/backup_test.go [hunk 23]
```diff
@@ -60,14 +54,107 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
...
+							ps[obj.Name] = obj.Spec.Template.Labels
+						}
+					} else if obj, ok := specObj.(*appsapi.StatefulSet); ok {
+						if contains(app_list, obj.Name) == true {
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:120]: same as above

File: tests/backup/backup_test.go [hunk 24]
```diff
@@ -60,14 +54,107 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
...
+							ps[obj.Name] = obj.Spec.Template.Labels
+						}
+					}
+				}
+			}
+		})
+		Step("Creating rules for backup", func() {
+			dash.Info("Creating pre rule for deployed apps")
+			pre_rule_status := createRuleForBackup("backup-pre-rule", "default", app_list, "pre", ps)
+			dash.VerifyFatal(pre_rule_status, true, fmt.Sprintf("Verifying pre rule for backup"))
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:130]: when there is no formatting required you can directly give the string

File: tests/backup/backup_test.go [hunk 25]
```diff
@@ -60,14 +54,107 @@ func TearDownBackupRestore(bkpNamespaces []string, restoreNamespaces []string) {
...
+				CreateBackupLocation(provider, backup_location_name, BackupLocationUID, CredName, CloudCredUID, bucketName, orgID)
+			}
+		})
+	})
+	JustAfterEach(func() {
+
+		teardown_status := teardownForTestcase(contexts, providers, CloudCredUID_list)
+		dash.VerifyFatal(teardown_status, true, fmt.Sprintf("Testcase teardown status"))
+		defer dash.TestCaseEnd()
+	})
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:156]: defer should be first statement.
defer makes sure the statement /function is executed always

File: tests/backup/common.go [hunk 26]
```diff
@@ -0,0 +1,273 @@
...
+	if err != nil {
+		log.Errorf("Unable to fetch pxcentral-post-install-hook pod from backup namespace\n Error : [%v]\n",
+			err)
+		return false
+	}
+	for _, pod := range pods.Items {
+		log.Infof("Checking if the pxcentral-post-install-hook pod is in Completed state or not")
+		bkp_pod, err := core.Instance().GetPodByName(pod.GetName(), ns)
+		if err != nil {
+			log.Errorf("An Error Occured while getting the pxcentral-post-install-hook pod details")
```
Comment (lsrinivas-pure) [inline tests/backup/common.go:59]: print the error here,
or return the error and handle it in calling function

File: tests/backup/common.go [hunk 27]
```diff
@@ -0,0 +1,273 @@
...
+		matched, _ := regexp.MatchString(post_install_hook_pod, pod.GetName())
+		if !matched {
+			equal, _ := regexp.MatchString(quick_maintenance_pod, pod.GetName())
+			equal1, _ := regexp.MatchString(full_maintenance_pod, pod.GetName())
+			if !(equal || equal1) {
+				log.Infof("Checking if all the containers are up or not")
+				res := core.Instance().IsPodRunning(pod)
+				if !res {
+					log.Errorf("All the containers are not Up")
+					return false
```
Comment (lsrinivas-pure) [inline tests/backup/common.go:86]: log the pod name here

File: tests/backup/common.go [hunk 28]
```diff
@@ -0,0 +1,273 @@
...
+				}
+				err = core.Instance().ValidatePod(&pod, defaultTimeout, defaultTimeout)
+				log.Warnf("ERR is %s", err)
```
Comment (lsrinivas-pure) [inline tests/backup/common.go:89]: you can delete this as you are printing the err in next line

File: tests/backup/common.go [hunk 29]
```diff
@@ -0,0 +1,273 @@
...
+	// User should keep updating the below 3 datas
+	pre_rule_app   = []string{"cassandra", "postgres"}
+	post_rule_app  = []string{"cassandra"}
+	app_parameters = map[string]map[string]string{
+		"cassandra": {"pre_action_list": "nodetool flush -- keyspace1;", "post_action_list": "nodetool verify -- keyspace1;", "background": "false", "run_in_single_pod": "false"},
+		"postgres":  {"pre_action_list": "PGPASSWORD=$POSTGRES_PASSWORD; psql -U '$POSTGRES_USER' -c 'CHECKPOINT';", "background": "false", "run_in_single_pod": "false"},
+	}
+)
+
+var log *logrus.Logger
```
Comment (lsrinivas-pure) [inline tests/backup/common.go:40]: where is the log and dash getting initialized

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

File: tests/backup/backup_basic_test.go (hunk 20 above)
Comment (snigdha-px) [reply tests/backup/backup_basic_test.go:60]: done!!

File: tests/backup/backup_test.go (hunk 21 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:101]: Done!!

File: tests/backup/backup_test.go (hunk 22 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:116]: Done!!

File: tests/backup/backup_test.go (hunk 23 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:120]: Done!!

File: tests/backup/backup_test.go (hunk 24 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:130]: Done!!

File: tests/backup/backup_test.go (hunk 25 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:156]: Done!!

File: tests/backup/common.go (hunk 27 above)
Comment (snigdha-px) [reply tests/backup/common.go:86]: Done!!

File: tests/backup/common.go (hunk 28 above)
Comment (snigdha-px) [reply tests/backup/common.go:89]: Done!!

File: tests/backup/common.go (hunk 26 above)
Comment (snigdha-px) [reply tests/backup/common.go:59]: Done!!

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

**Pull request merged**